pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/NoDictionaryOnHeapDictionaryJointRule.java

```diff
@@ -0,0 +1,245 @@
+      LOGGER.debug("bitCompressedDataSize {}", bitCompressedDataSize);
+      LOGGER.debug("dictionarySize {}", dictionarySize);
+      LOGGER.debug("numValuesPerEntry {}", numValuesPerEntry);
+
+      if (_inputManager.getTableType().equalsIgnoreCase(REALTIME)) {
+        //TODO: improve this estimation
+        noDictSize = // size of one segment flushed ith no dictionary
+            colDataSizeWithoutDictionary * numValuesPerEntry * _params.SEGMENT_FLUSH_TIME;
+        withDictSize = // size of one flushed segment with dictionary
+            dictionarySize + bitCompressedDataSize * numValuesPerEntry * _params.SEGMENT_FLUSH_TIME;
+      } else { // For hybrid or offline table, nodictionary follows the offline side
+        noDictSize = // size of all segments in one push  with no dictionary
+            colDataSizeWithoutDictionary * numValuesPerEntry * numRecordsPerPush;
+        withDictSize = // size of all segments in one push with dictionary
+            dictionarySize * dictionaryCoefficient(cardinality, numRecordsPerPush) * DEFAUlT_NUM_PARTITIONS
+                + bitCompressedDataSize * numValuesPerEntry * numRecordsPerPush;
+      }
+
+      double storageSaved = (double) (noDictSize - withDictSize) / noDictSize;
+      LOGGER.debug("colName {}, noDictSize {}, withDictSize{}, storageSaved{}", colName, noDictSize, withDictSize,
+          storageSaved);
+
+      if (storageSaved > _params.THRESHOLD_MIN_PERCENT_DICTIONARY_STORAGE_SVAE) {
+        noDictCols.remove(colName);
+      }
+    }
+
+    // Add the no dictionary cols to config
+    _outputManager.getIndexConfig().getNoDictionaryColumns().addAll(noDictCols);
+
+    //**********On heap dictionary recommendation*******/
+    if (_inputManager.getQps() > _params.THRESHOLD_MIN_QPS_ON_HEAP) { // QPS > THRESHOLD_MIN_QPS_ON_HEAP
+      for (String colName : _inputManager.getColNameToIntMap().keySet()) {
+        if (!_outputManager.getIndexConfig().getNoDictionaryColumns().contains(colName)) //exclude no dictionary column
+        {
+          long dictionarySize = _inputManager.getDictionarySize(colName);
+          int colId = _inputManager.colNameToInt(colName);
+          double filterGroupByFreq = filterGroupByWeights[colId] / totalWeight.get();
+          if (filterGroupByFreq > _params.THRESHOLD_MIN_FILTER_FREQ_ON_HEAP  //frequently used in filter/group by
+              && dictionarySize < _params.THRESHOLD_MAX_DICTIONARY_SIZE_ON_HEAP) { // memory foot print < threshold
+            _outputManager.getIndexConfig().getOnHeapDictionaryColumns().add(colName);
+          }
+        }
+      }
+    }
+  }
+
+  private double dictionaryCoefficient(double cardinality, long numRecordsPerPush) {
+    return 1 - min(max(DEFAULT_DICT_COEFF_A * Math.log(DEFAULT_DICT_COEFF_B * cardinality / numRecordsPerPush),
+        DEFAULT_DICT_LOWER), DEFAULT_DICT_UPPER);
+  }
+
+  public void parseQuery(String queryString, double weight, double[] filterGroupByWeights, double[] selectionWeights) {
```

Comment:
General question - Would it be possible to parse the query exactly once before the execution of first rule begins. Right now, it seems like as the rules are fired in order, each rule will parse the input query set? Even though the algorithm of each rule is different, is it possible to parse once and extract all the common info needed by all the rules?

Reply:
resolved - the code will parse only once